### PR TITLE
[wip] OIDC Discovery override

### PR DESCRIPTION
**Pull request by mikz**
Override configuration from OIDC Discovery.
For example "public_key", because APIcast does not support JWK yet.

### CODE REVIEW

File: doc/forgerock.md [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+APIcast does OIDC Discovery only when downloading configuration from System. Passing configuration file from filesystem won't trigger that process.
+It is possible to mock System API by static files and a simple webserver:
+
+```shell
+gem install adsf
+cd examples/configuration/oidc/
+adsf
+```
+
+And edit `examples/configuration/oidc/admin/api/services/oidc/proxy/configs/production/latest.json` to provide the correct public key.
```
Comment (mikz) [inline doc/forgerock.md]: I used https://play.golang.org/p/mLpOxS-5Fy to convert the JWK to normal public key.

File: examples/configuration/oidc/admin/api/services/oidc/proxy/configs/production/latest.json [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        "hosts": [
+          "localhost",
+          "127.0.0.1"
+        ],
+        "backend": {
+          "endpoint": "http://127.0.0.1:8081",
+          "host": "echo"
+        },
+        "policy_chain": [
+          { "name": "apicast.policy.oidc_default_configuiration", "configuration": {
```
Comment (davidor) [inline examples/configuration/oidc/admin/api/services/oidc/proxy/configs/production/latest.json]: typo

File: gateway/src/apicast/oauth/oidc.lua [hunk 3]
```diff
@@ -96,9 +96,14 @@ local function parse_and_verify_token(self, jwt_token)
   if not self.alg_whitelist[jwt_obj.header.alg] then
     return jwt_obj, '[jwt] invalid alg'
   end
+
   -- TODO: this should be able to use DER format instead of PEM
   local pubkey = format_public_key(self.config.public_key)
 
+  if not pubkey then
+    return jwt_obj, '[jwt] missing public key for signature verification'
```
Comment (davidor) [inline gateway/src/apicast/oauth/oidc.lua]: I think we should add a test for this.

File: t/apicast-policy-oidc_default_configuration.t [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+our $public_key_oneline = join('', grep { length($_) eq 64 } split(/\R/, $public_key));
+
+
+repeat_each(2);
+
+run_tests();
+
+__DATA__
+
+=== TEST 1: OIDC Default configuration policy can override private key
```
Comment (davidor) [inline t/apicast-policy-oidc_default_configuration.t]: public key

File: t/apicast-policy-oidc_default_configuration.t [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  }
+--- backend
+  location = /transactions/oauth_authrep.xml {
+    content_by_lua_block {
+      local expected = "provider_key=fookey&service_id=42&usage%5Bhits%5D=1&app_id=appid"
+      require('luassert').same(ngx.decode_args(expected), ngx.req.get_uri_args(0))
+    }
+  }
+--- request
+GET /t
```
Comment (davidor) [inline t/apicast-policy-oidc_default_configuration.t]: We had problems in the past with OIDC, it only worked for requests that were not cached or something like that. Should we make more requests just to make sure that the same does not happen again?

Comment (davidor) [reply t/apicast-policy-oidc_default_configuration.t]: :+1:

**Comment by mikz** [comment]
Closing in favour of #850

**Pull request closed without merging**